google/cloud/pubsub_v1/futures.py

```diff
@@ -140,47 +158,51 @@ def add_done_callback(self, callback):
         Returns:
             None
         """
-        if self.done():
-            return callback(self)
-        self._callbacks.append(callback)
+        with self._condition:
+            if self._state != concurrent.futures._base.FINISHED:
+                self._callbacks.append(callback)
+                return
+
+        return callback(self)
 
     def set_result(self, result):
         """Set the result of the future to the provided result.
 
         Args:
             result (Any): The result
         """
-        # Sanity check: A future can only complete once.
-        if self.done():
-            raise RuntimeError("set_result can only be called once.")
+        with self._condition:
+            # Sanity check: A future can only complete once.
+            if self._state == concurrent.futures._base.FINISHED:
+                raise RuntimeError("set_result can only be called once.")
+
+            self._result = result
+            self._state = concurrent.futures._base.FINISHED
+            self._completed.set()
 
-        # Set the result and trigger the future.
-        self._result = result
-        self._trigger()
+            for waiter in self._waiters:
+                waiter.add_result(self)
+
+        for callback in self._callbacks:
```

Comment:
Do we not call callbacks on exception (with an exception parameter?)

Reply:
We do, there's a similar loop at the end of the `set_exception()` method. Or did I misunderstand the question?

Reply:
Oops, was mistaken via the diff UI.